src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs

```diff
@@ -479,7 +480,12 @@ public static class Sse
         /// __m128 _mm_set1_ps (float a)
         ///   HELPER
         /// </summary>
-        public static Vector128<float> SetAllVector128(float value) => SetAllVector128(value);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<float> SetAllVector128(float value)
+        {
+            Vector128<float> vector = SetScalarVector128(value);
+            return Shuffle(vector, vector, 0);
+        }
```

Comment:
Could you try to optimize `SetAllVector128` via `Avx2.BroadcastScalarToVector128<T>`  to see whether the `if` block impacts inlining?
For example,
```csharp
        public static Vector128<float> SetAllVector128(float value)
        {
            if (Avx2.IsSupported)
            {
                return Avx2.BroadcastScalarToVector128<float>(value);
             }
             else
             {
                Vector128<float> vector = SetScalarVector128(value);
                return Shuffle(vector, vector, 0);
             }
        }
```

Reply:
Actually, the final solution should be:
```csharp
        public static Vector128<float> SetAllVector128(float value)
        {
             if (Avx2.IsSupported)
             {
                return Avx2.BroadcastScalarToVector128<float>(value);
             }
             else if (Sse41.IsSupported)
             {
                // use SSE4.1 insert
             }
             else
             {
                Vector128<float> vector = SetScalarVector128(value);
                return Shuffle(vector, vector, 0);
             }
        }
```
So I worry about that the inlining with the code size increasing.

Reply:
@AndyAyersMS, this type of code might fail the initial inlining check, but should be marked as inlineable after the dead code is dropped (all of the if-blocks not selected, since `IsSupported` is a constant), correct?

Reply:
I will test it shortly

Reply:
Not quite sure what you're asking -- the method likely will not be inlined without the aggressive inlining attribute. The inliner can't easily anticipate that the method body will shrink down once we've realized the calls therein expand to jit intrinsics with known values.

Generally we don't mark things as inlineable -- we reject them as possible candidates and possibly mark them as never inlinable to speed up future method jitting. The rejection happens in stages -- an initial pass in the importer and then a second pass in the inliner proper. Once a call is rejected it is never reconsidered, and (aside from the never inline stuff) we don't remember anything from past attempts, either when jitting this method or jitting any other.

Reply:
With brodcast the disasm is as follows:

```assembly
       C4E17A1005B6010000   vmovss   xmm0, dword ptr [reloc @RWD00]
       C4E17A1145AC         vmovss   dword ptr [rbp-54H], xmm0
       488D4DAC             lea      rcx, bword ptr [rbp-54H]
       C4E2791801           vbroadcastss xmm0, xmmword ptr [rcx]
```

We have again `vmovss` doing circles between stack and XMM but otherwise method is inlined and expanded as expected.

Reply:
Method implementation is:

```C#
        /// <summary>
        /// __m128 _mm_set1_ps (float a)
        ///   HELPER
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector128<float> SetAllVector128(float value)
        {
            if (Avx2.IsSupported)
            {
                 unsafe { return Avx.BroadcastScalarToVector128(&value); }
            }
            else
            {
                Vector128<float> vector = SetScalarVector128(value);
                return Shuffle(vector, vector, 0);
            }
        }
```

Reply:
No, we should use Avx2.BroadcastScalarToVector128 rather than Avx.

Reply:
@AndyAyersMS What do you think about having besides `AggressiveInlining` an `AlwaysInlining` value in `MethodImpOpt`. Would it be hard to get it implemented in jit?

Reply:
`AggressiveInlining` is already as aggressive as we can get -- if the jit can inline, it will. We are not holding anything back. The dial is already turned up to 11.

There are necessary limitations.

Some methods can't ever be inlined, or can't currently be inlined. And we pass on inlining in some rare situations -- for instance you can't use `AggressiveInlining` and recursion to create unboundedly large methods.

Reply:
> No, we should use Avx2.BroadcastScalarToVector128 rather than Avx.

`Avx2.BroadcastScalarToVector128` accepts `Vector128<float>` not float therefore signature in your example is invalid.

Reply:
> Avx2.BroadcastScalarToVector128 accepts Vector128<float> not float

Oh, I see. So that we also need `SetScalarVector128`, and we cannot use `Avx.BroadcastScalarToVector128` because storing arguments into memory and reading it again is really slow.
Looks like one more reason to implement these helpers in codgen.